src/vs/editor/contrib/inlineEdit/browser/inlineEditController.ts

```diff
@@ -0,0 +1,320 @@
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+
+export class InlineEditWidget implements IDisposable {
+	constructor(public readonly widget: GhostTextWidget, public readonly edit: IInlineEdit) { }
+
+	dispose(): void {
+		this.widget.dispose();
+	}
+}
+
+export class InlineEditController extends Disposable {
+	static ID = 'editor.contrib.inlineEditController';
+
+	public static readonly inlineEditVisibleKey = 'inlineEditVisible';
+	public static readonly inlineEditVisibleContext = new RawContextKey<boolean>(InlineEditController.inlineEditVisibleKey, false);
+	private _isVisibleContext = InlineEditController.inlineEditVisibleContext.bindTo(this.contextKeyService);
+
+	public static readonly cursorAtInlineEditKey = 'cursorAtInlineEdit';
+	public static readonly cursorAtInlineEditContext = new RawContextKey<boolean>(InlineEditController.cursorAtInlineEditKey, false);
+	private _isCursorAtInlineEditContext = InlineEditController.cursorAtInlineEditContext.bindTo(this.contextKeyService);
+
+	public static get(editor: ICodeEditor): InlineEditController | null {
+		return editor.getContribution<InlineEditController>(InlineEditController.ID);
+	}
+
+	private _currentEdit: ISettableObservable<InlineEditWidget | undefined> = this._register(disposableObservableValue(this, undefined));
+	private _currentRequestCts: CancellationTokenSource | undefined;
+
+	private _jumpBackPosition: Position | undefined;
+	private _isAccepting: boolean = false;
+
+	private readonly _enabled = observableFromEvent(this.editor.onDidChangeConfiguration, () => this.editor.getOption(EditorOption.inlineEdit).enabled);
+	private readonly _fontFamily = observableFromEvent(this.editor.onDidChangeConfiguration, () => this.editor.getOption(EditorOption.inlineEdit).fontFamily);
+	private readonly _backgroundColoring = observableFromEvent(this.editor.onDidChangeConfiguration, () => this.editor.getOption(EditorOption.inlineEdit).backgroundColoring);
+
+
+	constructor(
+		public readonly editor: ICodeEditor,
+		@IInstantiationService private readonly instantiationService: IInstantiationService,
+		@IContextKeyService private readonly contextKeyService: IContextKeyService,
+		@ILanguageFeaturesService private readonly languageFeaturesService: ILanguageFeaturesService,
+		@ICommandService private readonly _commandService: ICommandService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+	) {
+		super();
+
+		//Automatically request inline edit when the content was changed
+		//Cancel the previous request if there is one
+		//Remove the previous ghost text
+		this._register(editor.onDidChangeModelContent(async () => {
+			if (!this._enabled.get()) {
+				return;
+			}
+			await this.getInlineEdit(editor, true);
+		}));
+
+		//Check if the cursor is at the ghost text
+		this._register(editor.onDidChangeCursorPosition((e) => {
+			if (!this._enabled.get()) {
```

Comment:
Can you only subscribe to `onDidChangeCursorPosition` when enabled is true?
This feature will probably only be enabled for 0.001% of VS Code users and this code is called billions of times per day in the world.

You can do it like this:
```
const cursorPosition = observableFromEvent(editor.onDidChangeCursorPosition, () => editor.getPosition());
this._register(autorun(reader => {
	if (!this.enabled.read(reader)) { return; }

	const pos = cursorPosition.read(reader);
	if (pos) {
		this.checkCursorPosition(pos);
	}
}));
```

Reply:
I've made the change you suggested here, but I'm not 100% sure how it makes things better - I've thought `autorun` would be rerun if any observed `IObservable` has changed so changing `cursorPosition` (which happens on every cursor position change) would trigger this `autorun`... so it's exactly same behaviour as before?

Reply:
Oh, that makes sense! Thanks for explanation.